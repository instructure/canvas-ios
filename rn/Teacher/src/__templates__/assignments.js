//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const assignment: Template<Assignment> = template({
  id: '123456789',
  course_id: '987654321',
  name: 'learn to write code',
  description: 'a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
  created_at: '2017-03-17T19:15:25Z',
  updated_at: '2017-03-17T19:15:25Z',
  points_possible: 20,
  grading_type: 'points',
  assignment_group_id: '1376',
  peer_reviews: false,
  automatic_peer_reviews: false,
  position: 1,
  submission_types:
  [ 'online_text_entry',
    'online_url',
    'media_recording',
    'online_upload' ],
  needs_grading_count: 0,
  html_url: 'http://hostname.tld/whatever',
  overrides: [],
  grade_group_students_individually: false,
  group_category_id: null,
  has_overrides: false,
  only_visible_to_overrides: false,
  published: true,
  unpublishable: false,
  rubric_assessment: {},
  rubric_settings: null,
  rubric: null,
  anonymize_students: false,
})

export const assignmentGroup: Template<AssignmentGroup> = template({
  id: '1',
  name: 'Learn React Native',
  position: 1,
  assignments: [assignment({ assignment_group_id: '1' })],
  group_weight: 1,
  integration_data: {},
  sis_source_id: '',
})

export const assignmentDueDate: Template<AssignmentDate> = template({
  title: 'due',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})

export const assignmentOverride: Template<AssignmentOverride> = template({
  id: '1',
  assignment_id: '1',
  student_ids: [],
  group_id: '1',
  course_section_id: '1',
  title: 'override',
  all_day: false,
  all_date_date: '',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})
