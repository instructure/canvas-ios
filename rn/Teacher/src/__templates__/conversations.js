//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const conversation: Template<Conversation> = template({
  id: '123456789',
  subject: 'learn to write code',
  workflow_state: 'read',
  last_message: 'this is the worst class ever',
  last_message_at: '2037-06-01T05:59:00Z',
  last_authored_message: 'this is the worst class ever',
  last_authored_message_at: '2037-06-01T05:59:00Z',
  participants: [
    {
      id: '1234',
      name: 'artorias of the abyss',
    },
    {
      id: '6789',
      name: 'solaire of astora',
    },
  ],
  avatar_url: 'http://www.fillmurray.com/200/300',
  context_name: 'Test Class',
  context_code: 'course_1',
  audience_contexts: {},
  audience: [],
  message_count: 2,
  private: false,
  properties: null,
  starred: false,
  subscribed: false,
  visible: true,
})

export const conversationMessage: Template<ConversationMessage> = template({
  id: '123456789',
  body: 'this is neat',
  author_id: '1234',
  created_at: '2037-06-01T05:59:00Z',
  attachments: [
    {
      id: '1',
      display_name: 'something.pdf',
      mime_class: 'pdf',
    },
  ],
  generated: false,
})
