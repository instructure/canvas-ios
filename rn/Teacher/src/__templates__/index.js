//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export * from '../redux/__templates__/app-state'
export * from './account-notification'
export * from './addressBook'
export * from './api'
export * from './assignments'
export * from './attachment'
export * from './conversations'
export * from './course'
export * from './discussion'
export * from './enrollments'
export * from './error'
export * from './external-tool'
export * from './file'
export * from './grading-periods'
export * from './group'
export * from './helm'
export * from './launch-definitions'
export * from './mediaComment'
export * from './quiz'
export * from './quizSubmission'
export * from './role'
export * from './rubric'
export * from './section'
export * from './session'
export * from './submissions'
export * from './tab'
export * from './users'
export * from './help-links'
export * from './dashboard'
export * from '../modules/assignee-picker/__template__/Assignee'
