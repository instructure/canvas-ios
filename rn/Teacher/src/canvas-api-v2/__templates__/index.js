// @flow
export * from './submission'
export * from './user'
export * from './assignment'
export * from './course'
export * from './group'
export * from './section'
export * from './groupSet'

// QUERIES
export * from '../queries/__templates__/SubmissionList'
export * from '../queries/__templates__/ContextCard'
