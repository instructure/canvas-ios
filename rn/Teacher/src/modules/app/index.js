//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

export type AppId = 'student' | 'teacher' | 'parent'
export type App = {
  appId: AppId,
  filterCourse: (course: Course) => boolean,
}

const teacher = {
  appId: 'teacher',
  filterCourse: (course: Course): boolean => {
    if (course.access_restricted_by_date) return false
    const enrollments = course.enrollments
    if (!enrollments) return false
    return enrollments.some((e) =>
      [
        'teacher',
        'teacherenrollment',
        'designer',
        'ta',
      ].includes(e.type.toLowerCase())
    )
  },
}

const student = {
  appId: 'student',
  filterCourse: (course: Course): boolean => !course.access_restricted_by_date,
}

const parent = {
  appId: 'parent',
  filterCourse: (course: Course): boolean => true,
}

let current: App = teacher

const app = {
  setCurrentApp: (appId: AppId): void => {
    switch (appId) {
      case 'student':
        current = student
        break
      case 'teacher':
        current = teacher
        break
      case 'parent':
        current = parent
        break
    }
  },
  current: (): App => current,
}

export function isTeacher (): boolean {
  return app.current().appId === 'teacher'
}
export function isStudent (): boolean {
  return app.current().appId === 'student'
}

export function isParent (): boolean {
  return app.current().appId === 'parent'
}

export default app
