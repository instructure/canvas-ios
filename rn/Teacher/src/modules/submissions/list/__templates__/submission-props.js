//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
