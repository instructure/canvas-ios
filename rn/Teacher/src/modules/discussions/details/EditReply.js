//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import i18n from 'format-message'
import { colors } from '../../../common/stylesheet'
import Screen from '../../../routing/Screen'
import RichTextEditor from '../../../common/components/rich-text-editor/RichTextEditor'
import Actions from './actions'
import { isTeacher } from '../../app'
import Images from '../../../images'
import {
  ActivityIndicator,
  View,
  LayoutAnimation,
  processColor,
  NativeModules,
  StyleSheet,
} from 'react-native'

type OwnProps = {
  discussionID: string,
  context: CanvasContext,
  contextID: string,
  indexPath?: number[],
  entryID?: ?string,
  lastReplyAt: string,
  message: string,
  permissions: DiscussionPermissions,
  isEdit?: boolean,
}

type State = {
  pending: number,
  error?: ?string,
}

type Props = OwnProps & typeof Actions & NavigationProps & State

export class EditReply extends React.Component<Props, any> {
  state: any = {
    pending: false,
    attachment: null,
  }
  editor: ?RichTextEditor

  render () {
    const permissions = this.props.permissions
    let message = this.props.message || ''
    const { isEdit } = this.props
    return (
      <Screen
        title={isEdit ? i18n('Edit') : i18n('Reply')}
        navBarTitleColor={colors.textDarkest}
        navBarButtonColor={colors.linkColor}
        rightBarButtons={[
          {
            title: i18n('Done'),
            style: 'done',
            testID: 'DiscussionEditReply.doneButton',
            action: this._actionDonePressed,
          },
          permissions && permissions.attach && {
            image: Images.paperclip,
            testID: 'DiscussionEditReply.attachmentButton',
            action: this.addAttachment,
            accessibilityLabel: i18n('Edit attachment ({count})', { count: this.state.attachment ? '1' : i18n('none') }),
            badge: this.state.attachment && {
              text: '1',
              backgroundColor: processColor(colors.backgroundInfo),
              textColor: processColor(colors.white),
            },
          },
        ]}
        dismissButtonTitle={i18n('Cancel')}
      >
        <View style={{ flex: 1 }}>
          <RichTextEditor
            ref={(r) => { this.editor = r }}
            onChangeValue={this._valueChanged('message')}
            defaultValue={message}
            showToolbar='always'
            scrollEnabled={true}
            placeholder={i18n('Message')}
            focusOnLoad={true}
            navigator={this.props.navigator}
            attachmentUploadPath={isTeacher() ? `/${this.props.context}/${this.props.contextID}/files` : '/users/self/files'}
            context={this.props.context}
            contextID={this.props.contextID}
          />
          { this.state.pending &&
            <ActivityIndicator style={StyleSheet.absoluteFill} size='large' />
          }
        </View>
      </Screen>
    )
  }

  componentWillUnmount () {
    this.props.deletePendingReplies(this.props.discussionID)
  }

  _actionDonePressed = async () => {
    const message = this.editor && await this.editor.getHTML()
    const params = {
      message,
      attachment: this.state.attachment,
    }
    this.setState({ pending: true })
    const isNew = !this.props.isEdit
    try {
      if (isNew) {
        await this.props.createEntry(this.props.context, this.props.contextID, this.props.discussionID, this.props.entryID, params, this.props.indexPath, this.props.lastReplyAt).payload.promise
      } else {
        await this.props.editEntry(this.props.context, this.props.contextID, this.props.discussionID, this.props.entryID, params, this.props.indexPath).payload.promise
      }
      this.props.refreshDiscussionEntries(this.props.context, this.props.contextID, this.props.discussionID, true)
      await this.props.navigator.dismiss()
      if (isNew) {
        NativeModules.AppStoreReview.handleSuccessfulSubmit()
        NativeModules.ModuleItemsProgress.contributedDiscussion(this.props.contextID, this.props.discussionID)
      }
    } catch (e) {
      console.warn(e)
      this.setState({ pending: false })
    }
  }

  _valueChanged (property: string, transformer?: any, animated?: boolean = true): Function {
    return (value) => {
      if (transformer) { value = transformer(value) }
      this._valuesChanged({ [property]: value }, animated)
    }
  }

  _valuesChanged (values: Object, animated?: boolean) {
    if (animated) {
      LayoutAnimation.easeInEaseOut()
    }
    this.setState({ ...values })
  }

  addAttachment = () => {
    this.props.navigator.show('/attachments', { modal: true }, {
      attachments: this.state.attachment ? [this.state.attachment] : [],
      maxAllowed: 1,
      storageOptions: {
        uploadPath: null,
      },
      onComplete: this._valueChanged('attachment', (as) => as[0]),
    })
  }
}

export function mapStateToProps ({ entities }: AppState, { context, contextID, discussionID }: OwnProps): State {
  let pending = 0
  let error = null

  if (entities.discussions[discussionID] &&
      entities.discussions[discussionID].replies &&
      entities.discussions[discussionID].replies.new) {
    pending = entities.discussions[discussionID].replies.new.pending
    error = entities.discussions[discussionID].replies.new.error
  }
  if (!pending && !error &&
      entities.discussions[discussionID] &&
      entities.discussions[discussionID].replies &&
      entities.discussions[discussionID].replies.edit) {
    pending = entities.discussions[discussionID].replies.edit.pending
    error = entities.discussions[discussionID].replies.edit.error
  }

  return {
    pending,
    error,
  }
}

let Connected = connect(mapStateToProps, Actions)(EditReply)
export default (Connected: Component<Props, any>)
