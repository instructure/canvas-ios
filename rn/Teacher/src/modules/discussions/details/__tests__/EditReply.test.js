//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import { shallow } from 'enzyme'
import React from 'react'
import { EditReply, mapStateToProps } from '../EditReply'
import { NativeModules } from 'react-native'
import app from '../../../app'
import * as template from '../../../../__templates__'

jest
  .mock('react-native/Libraries/LayoutAnimation/LayoutAnimation', () => ({
    easeInEaseOut: jest.fn(),
    Types: {
      easeInEaseOut: jest.fn(),
      spring: jest.fn(),
    },
    Properties: {
      opacity: 1,
    },
  }))

describe('EditReply', () => {
  let defaultProps
  beforeEach(() => {
    jest.clearAllMocks()
    defaultProps = {
      discussion: template.discussion({ id: '1' }),
      navigator: template.navigator(),
      discussionID: '1',
      context: 'courses',
      contextID: '1',
      entryID: '1',
      course: template.course({ id: 1 }),
      indexPath: [],
      deletePendingReplies: jest.fn(),
      lastReplyAt: (new Date()).toISOString(),
      permissions: {
        attach: true,
        delete: true,
        reply: true,
        update: true,
      },
    }
    app.setCurrentApp('teacher')
  })

  it('renders title correctly', () => {
    let tree = shallow(<EditReply {...defaultProps} />)
    expect(tree.find('Screen').prop('title')).toEqual('Reply')

    tree = shallow(<EditReply {...defaultProps} isEdit />)
    expect(tree.find('Screen').prop('title')).toEqual('Edit')
  })

  it('deletes pending replies on unmount', () => {
    let tree = shallow(
      <EditReply {...defaultProps} />
    )
    tree.unmount()
    expect(defaultProps.deletePendingReplies).toHaveBeenCalledWith(defaultProps.discussionID)
  })

  it('dismisses modal activity upon save error', async () => {
    let promise = Promise.reject('error')
    let postReply = jest.fn(() => ({ payload: { promise } }))
    let tree = shallow(
      <EditReply
        {...defaultProps}
        createEntry={postReply}
      />
    )

    const doneButton = tree.find('Screen').prop('rightBarButtons')[0]
    doneButton.action()
    expect(tree.find('ActivityIndicator').exists()).toEqual(true)
    try { await promise } catch (_) {}
    expect(tree.find('ActivityIndicator').exists()).toEqual(false)
  })

  it('dismisses modal after reply updates', async () => {
    let postReply = jest.fn(() => ({ payload: { promise: Promise.resolve() } }))
    let refresh = jest.fn(() => ({ payload: { promise: Promise.resolve() } }))
    let tree = shallow(
      <EditReply
        {...defaultProps}
        createEntry={postReply}
        refreshDiscussionEntries={refresh}
      />
    )
    const doneButton = tree.find('Screen').prop('rightBarButtons')[0]
    await doneButton.action()
    expect(postReply).toHaveBeenCalled()
    expect(defaultProps.navigator.dismiss).toHaveBeenCalled()
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit).toHaveBeenCalled()
    expect(NativeModules.ModuleItemsProgress.contributedDiscussion).toHaveBeenCalledWith(defaultProps.contextID, defaultProps.discussionID)
  })

  it('sets message placeholder', () => {
    let tree = shallow(<EditReply {...defaultProps} />)
    let textEditor = tree.find('RichTextEditor')
    expect(textEditor.prop('placeholder')).toEqual('Message')
  })

  it('enters text and posts reply', async () => {
    let refresh = jest.fn(() => ({ payload: { promise: Promise.resolve() } }))
    let component = shallow(<EditReply {...defaultProps} refreshDiscussionEntries={refresh} />)
    let postReply = jest.fn(() => ({ payload: { promise: Promise.resolve() } }))
    component.setProps({ createEntry: postReply })
    let textEditor = component.find('RichTextEditor')
    let message = 'not empty'
    textEditor.getElement().ref({ getHTML: jest.fn(() => Promise.resolve(message)) })
    const doneButton = component.prop('rightBarButtons')[0]
    await doneButton.action()
    expect(postReply).toBeCalledWith(defaultProps.context, defaultProps.contextID, defaultProps.discussionID, defaultProps.entryID, { attachment: null, message }, [], defaultProps.lastReplyAt)
    expect(defaultProps.navigator.dismiss).toHaveBeenCalled()
  })

  it('edits an existing reply', async () => {
    let editProps = {
      ...defaultProps,
      message: 'default message',
      isEdit: true,
      refreshDiscussionEntries: jest.fn(() => ({ payload: { promise: Promise.resolve() } })),
    }
    let component = shallow(<EditReply {...editProps} />)
    let editEntry = jest.fn(() => ({ payload: { promise: Promise.resolve() } }))
    component.setProps({ editEntry })
    let textEditor = component.find('RichTextEditor')
    let message = 'edited message'
    textEditor.getElement().ref({ getHTML: jest.fn(() => Promise.resolve(message)) })
    const doneButton = component.prop('rightBarButtons')[0]
    await doneButton.action()
    expect(editEntry).toBeCalledWith(defaultProps.context, defaultProps.contextID, editProps.discussionID, editProps.entryID, { attachment: null, message }, [])
    expect(defaultProps.navigator.dismiss).toHaveBeenCalled()
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit).not.toHaveBeenCalled()
    expect(NativeModules.ModuleItemsProgress.contributedDiscussion).not.toHaveBeenCalled()
  })

  it('cannot add an attachment if it does not have permission', () => {
    defaultProps.permissions.attach = false
    const tree = shallow(<EditReply {...defaultProps} />)
    const attach = tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'DiscussionEditReply.attachmentButton')
    expect(attach).toBeUndefined()
  })

  it('can add an attachment if it has permission', () => {
    defaultProps.navigator = template.navigator({ show: jest.fn() })
    defaultProps.permissions.attach = true
    const tree = shallow(<EditReply {...defaultProps} />)
    const attach = tree.find('Screen').prop('rightBarButtons')
      .find(({ testID }) => testID === 'DiscussionEditReply.attachmentButton')
    expect(attach).toBeDefined()
    attach.action()
    expect(defaultProps.navigator.show).toHaveBeenCalledWith(
      '/attachments',
      { modal: true },
      {
        attachments: [],
        maxAllowed: 1,
        storageOptions: {
          uploadPath: null,
        },
        onComplete: expect.any(Function),
      }
    )
    const attachment = template.attachment()
    defaultProps.navigator.show.mock.calls[0][2].onComplete([ attachment ])
    expect(tree.state('attachment')).toBe(attachment)
  })

  it('uses course files for rce media embeds in teacher', () => {
    app.setCurrentApp('teacher')
    defaultProps.context = 'courses'
    defaultProps.contextID = '1'
    const screen = shallow(<EditReply {...defaultProps} />)
    expect(screen.find('RichTextEditor').prop('attachmentUploadPath')).toEqual('/courses/1/files')
  })

  it('uses user files for rce media embeds in student', () => {
    app.setCurrentApp('student')
    const screen = shallow(<EditReply {...defaultProps} />)
    expect(screen.find('RichTextEditor').prop('attachmentUploadPath')).toEqual('/users/self/files')
  })
})

describe('MapStateToProps', () => {
  let lastReplyAt = (new Date(0)).toISOString()
  test('maps default data for new reply', () => {
    const state: AppState = template.appState({
      entities: {
        discussions: {

        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', indexPath: [], lastReplyAt })
    ).toMatchObject({
      pending: 0,
      error: null,
    })
  })

  test('finds the correct data for new reply', () => {
    const state: AppState = template.appState({
      entities: {
        ...template.appState().entities,
        discussions: {
          '1': {
            pending: 0,
            error: null,
            replies: {
              pending: 0,
              error: null,
              refs: [],
              new: {
                pending: 14,
                error: 'Map this error',
              },
            },
          },
        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', lastReplyAt })
    ).toMatchObject({
      pending: 14,
      error: 'Map this error',
    })
  })

  test('finds the correct data for editing a reply', () => {
    const state: AppState = template.appState({
      entities: {
        ...template.appState().entities,
        discussions: {
          '1': {
            pending: 0,
            error: null,
            replies: {
              pending: 0,
              error: null,
              refs: [],
              edit: {
                pending: 14,
                error: 'Map this error',
              },
            },
          },
        },
      },
    })
    expect(
      mapStateToProps(state, { courseID: '1', discussionID: '1', lastReplyAt })
    ).toMatchObject({
      pending: 14,
      error: 'Map this error',
    })
  })
})
