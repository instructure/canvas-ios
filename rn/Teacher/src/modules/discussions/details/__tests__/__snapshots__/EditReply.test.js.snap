// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditReply dismisses modal activity upon save error 1`] = `
<Screen
  dismissButtonTitle="Cancel"
  navBarButtonColor="#008EE2"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
      Object {
        "accessibilityLabel": "Edit attachment (none)",
        "action": [Function],
        "badge": null,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "edit-discussion-reply.attachment-btn",
      },
    ]
  }
  title="Reply"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      attachmentUploadPath="/courses/1/files"
      defaultValue=""
      focusOnLoad={true}
      navigator={
        Object {
          "dismiss": [MockFunction],
          "dismissAllModals": [MockFunction],
          "isModal": false,
          "pop": [MockFunction],
          "replace": [MockFunction],
          "show": [MockFunction],
          "showWebView": [MockFunction],
          "traitCollection": [MockFunction],
        }
      }
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;

exports[`EditReply renders 1`] = `
<Screen
  dismissButtonTitle="Cancel"
  navBarButtonColor="#008EE2"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
      Object {
        "accessibilityLabel": "Edit attachment (none)",
        "action": [Function],
        "badge": null,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "edit-discussion-reply.attachment-btn",
      },
    ]
  }
  title="Reply"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      attachmentUploadPath="/courses/1/files"
      defaultValue=""
      focusOnLoad={true}
      navigator={
        Object {
          "dismiss": [MockFunction],
          "dismissAllModals": [MockFunction],
          "isModal": false,
          "pop": [MockFunction],
          "replace": [MockFunction],
          "show": [MockFunction],
          "showWebView": [MockFunction],
          "traitCollection": [MockFunction],
        }
      }
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;

exports[`EditReply renders title correctly when editing 1`] = `
<Screen
  dismissButtonTitle="Cancel"
  navBarButtonColor="#008EE2"
  navBarTitleColor="#2D3B45"
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "style": "done",
        "testID": "edit-discussion-reply.done-btn",
        "title": "Done",
      },
      Object {
        "accessibilityLabel": "Edit attachment (none)",
        "action": [Function],
        "badge": null,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "edit-discussion-reply.attachment-btn",
      },
    ]
  }
  title="Edit"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      transparent={true}
      visible={false}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0,0,0,0.6)",
                "borderRadius": 12,
              },
              Object {
                "backgroundColor": "rgba(0,0,0,0.8)",
              },
              Object {
                "height": 170,
                "width": 170,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#fff",
                      "fontFamily": ".SFUIDisplay-semibold",
                      "fontSize": 24,
                    },
                    Array [
                      Object {
                        "alignItems": "center",
                        "color": "#fff",
                        "textAlign": "center",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Saving
              </Text>
            </View>
          </View>
          <ActivityIndicator
            animating={true}
            color="#fff"
            hidesWhenStopped={true}
            size="large"
            style={
              Object {
                "flex": 1,
                "marginTop": -40,
              }
            }
          />
        </View>
      </View>
    </Modal>
    <RichTextEditor
      attachmentUploadPath="/courses/1/files"
      defaultValue=""
      focusOnLoad={true}
      navigator={
        Object {
          "dismiss": [MockFunction],
          "dismissAllModals": [MockFunction],
          "isModal": false,
          "pop": [MockFunction],
          "replace": [MockFunction],
          "show": [MockFunction],
          "showWebView": [MockFunction],
          "traitCollection": [MockFunction],
        }
      }
      onChangeValue={[Function]}
      placeholder="Message"
      scrollEnabled={true}
      showToolbar="always"
    />
  </View>
</Screen>
`;
